org.metaborg.spoofax.core/src/main/java/org/metaborg/spoofax/core/analysis/constraint/AbstractConstraintAnalyzer.java

```diff
@@ -275,4 +349,92 @@ private IMessage message(ITerm originatingTerm, IMessageInfo messageInfo, Messag
         };
     }
 
+    protected void flowspecDemoOutput(C context, final IControlFlowGraph<CFGNode> controlFlowGraph) {
+        logger.debug("Outputting FlowSpec demo output file");
+        FileObject file = this.resource("target/flowspec-demo-output.aterm", context);
+        try (PrintWriter out = new PrintWriter(file.getContent().getOutputStream())) {
+            out.println("(CFG([");
+            boolean first = true;
+            for (Map.Entry<CFGNode, CFGNode> e : controlFlowGraph.getDirectEdges().entrySet()) {
+                if (!first) {
+                    out.print(", ");
+                }
+                first = false;
+                out.println("(\"" + e.getKey() + "\", \"" + e.getValue() + "\")");
+            }
+            out.println("]),");
+            out.println("Properties([");
+            first = true;
+            for (Map.Entry<Tuple2<CFGNode, String>, ITerm> e : controlFlowGraph.getProperties().entrySet()) {
+                if(!first) {
+                    out.print(", ");
+                }
+                first = false;
+                out.println(e.getKey()._2() + "(\"" + e.getKey()._1() + "\", " + e.getValue().toString() +")");
+            }
+            out.println("]))");
+        } catch (FileSystemException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public static void flowspecCopyTFAppls(IControlFlowGraph<CFGNode> controlFlowGraph, IProperties.Immutable<TermIndex, ITerm, ITerm> astProperties) {
+        ControlFlowGraph<CFGNode> cfg = (ControlFlowGraph<CFGNode>) controlFlowGraph;
+        astProperties.stream().forEach(tuple -> {
+            ITerm key = tuple._2();
+            if(key instanceof IApplTerm) {
+                IApplTerm keyAppl = (IApplTerm) key;
+                if(keyAppl.getOp().equals("TF") && keyAppl.getArity() == 1) {
+                    ITerm propertyName = keyAppl.getArgs().get(0);
+                    assert propertyName instanceof IStringTerm : "Property name in TF property is not a string";
+                    TermIndex index = tuple._1();
+                    ITerm tfApplTerm = tuple._3();
+
+                    cfg.addTFAppl(index, ((IStringTerm) propertyName).getValue(), matchTFAppl(tfApplTerm));
+                }
+            }
+        });
+    }
+
+    private static TransferFunctionAppl matchTFAppl(ITerm pairTerm) {
```

Comment:
Same as before, why not use matching functions here, instead of this explict testing/casting etc?